Adding RBAC syntax tests.

vzakaznikov opened the pull request:
Adding RBAC syntax tests in TestFlows.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Adding RBAC syntax tests in TestFlows.

Detailed description / Documentation draft:
Adding RBAC syntax tests in TestFlows.

vzakaznikov committed:
* Adding rbac module
* Adding rbac syntax tests

vzakaznikov committed:
* Fixing issue with configs missings in the config.d folder
* Fixing import in alter_quota.py

vzakaznikov commented:
@vitlibar, could you please review this PR? This PR includes syntax tests for RBAC. Thanks!

vitlibar commented:
> @vitlibar, could you please review this PR? This PR includes syntax tests for RBAC. Thanks!

Yes, I'll look at it carefully.

vzakaznikov committed:
Adding execution of the `rbac.regression` module.

vzakaznikov commented:
@vitlibar, any chance to get this merged this week? We have other PR's almost ready to be pushed.

vitlibar commented on tests/testflows/rbac/regression.py:
```diff
@@ -0,0 +1,61 @@
+    "syntax/create role/I create role if not exists, role does exist":
+        [(Fail, issue_12507)],
+    "syntax/create settings profile/I create settings profile if not exists, profile does exist":
+        [(Fail, issue_12507)],
```
Will the testflows checks in out CI fail after I fix https://github.com/ClickHouse/ClickHouse/issues/12507 ?

vzakaznikov replied:
No, right now these are hard-coded xfails and they will need to be removed once you fix #12507.

vitlibar commented on tests/testflows/rbac/requirements/requirements.py:
```diff
@@ -0,0 +1,6132 @@
+# These requirements were auto generated
```
This file was generated from some file named requirements.md as far as I understood. Could you add that source file to our repository as well?

vzakaznikov replied:
Ok, I will ask @alex-zaitsev if we can release the source for our internal RBAC requirements.

vitlibar commented on tests/testflows/rbac/requirements/requirements.py:
```diff
@@ -0,0 +1,6132 @@
+# from software requirements specification (SRS)
+# document by TestFlows v1.6.200629.1155222.
+# Do not edit by hand but re-generate instead
+# using 'tfs requirements generate' command.
```
It would be nice if you make your generation tool write here the path to the `requirements.md` file.

vzakaznikov replied:
Noted. But not sure that if it would be that useful as it would just be a path to somewhere on my machine. I think we if add the source for the requirements to this folder it will be self explanatory.

vitlibar commented on tests/testflows/rbac/requirements/requirements.py:
```diff
@@ -0,0 +1,6132 @@
+# using 'tfs requirements generate' command.
+from testflows.core import Requirement
+
+RQ_SRS_006_RBAC = Requirement(
```
What does the `006` mean?

vzakaznikov replied:
This our internal SRS `id`. In this case the `id` for this SRS is `SRS-006`.

vitlibar replied:
I mean it's just a number from `requirements.md`, it doesn't mean anything special, isn't it?

vzakaznikov replied:
Yes, it just a number.

vitlibar commented:
Could you add to README.md information about how to run locally a specific test, not all of them. It's more likely that one need to run locally one specific test only.

vitlibar commented on tests/testflows/rbac/tests/syntax/alter_quota.py:
```diff
@@ -0,0 +1,206 @@
+                RQ_SRS_006_RBAC_Quota_Alter_Rename("1.0")]):
+            new_quota = "quota1"
+
+            try:
```
What's the difference between `Given` and `When`? It looks like in both cases you do something and you check something.

vzakaznikov replied:
You can find more information in the TestFlows handbook. Here are a few links: https://testflows.com/handbook/#Given, and https://testflows.com/handbook/#When.

vitlibar commented on tests/testflows/rbac/tests/syntax/alter_row_policy.py:
```diff
@@ -0,0 +1,244 @@
+import rbac.tests.errors as errors
+
+@TestFeature
+@Name("alter row policy")
```
I am not very good at python, could you help me understand what exactly the line like `@Name("alter row policy")` does?

vzakaznikov replied:
It is calling a decorator. A decorator is a function or an object that decorates some function. In this case these decorators decorate a function that implements the body of the test. So in this case we have

```python
@TestFeature
@Name("alter row policy")
```

where `@TestFeature` says that we are "defining" a `Feature` and the `@Name` will set the name of the `Feature` to `alter row policy`. Here are some references:

* https://testflows.com/handbook/#Decorated
* https://testflows.com/handbook/#Test-Definition-Classes
* https://www.programiz.com/python-programming/decorator

vitlibar replied:
Thanks for the explanation.

vitlibar commented on tests/testflows/rbac/tests/syntax/alter_row_policy.py:
```diff
@@ -0,0 +1,244 @@
+                    node.query(f"ALTER ROW POLICY policy8a ON default.foo TO {role}", exitcode=exitcode, message=message)
+            del role
+
+        with Scenario("I alter row policy to assign to all excpet role that does not exist, throws exception", flags=TE, requirements=[
```
What do the flags `TE` mean?

vzakaznikov replied:
It means that we are marking this test as `test-to-end` (https://testflows.com/handbook/#TE) and it means that
the test flow will not abort if this test does not pass. By default test flow would be aborted on fail (`first-fail`).

vitlibar commented on tests/testflows/rbac/tests/syntax/alter_settings_profile.py:
```diff
@@ -0,0 +1,232 @@
+            with When("I short form alter settings profile"):
+                node.query("ALTER PROFILE profile0")
+
+        with Scenario("I alter settings profile that does not exist, throws exception", flags=TE, requirements=[RQ_SRS_006_RBAC_SettingsProfile_Alter("1.0")]):
```
What will happen if we edit the requirements.md and remove some requirements there which are still referenced in our tests in *.py files?

vzakaznikov replied:
Unless you regenerate the `requirements.py` using the updated `requirements.md` then nothing will happen. If we do regenerate and get a new `requirements.py` that does not define requirements that are referenced in tests then you will get an `AttributeError` or something like it when you try to run the tests.

vzakaznikov commented:
> Could you add to README.md information about how to run locally a specific test, not all of them. It's more likely that one need to run locally one specific test only.

Yes. I will add it. In general, this information is covered in https://testflows.com/handbook/#Filtering.

vzakaznikov committed:
* Adding requirements markdown source.
* Updating README.md to include an example of how to run selected tests.

vzakaznikov committed:
Updating README.md.

vzakaznikov commented:
@vitlibar, I've added the Markdown source for the [requirements](https://github.com/ClickHouse/ClickHouse/blob/a03d1d138126548883f5acc636abf2a21e17d94f/tests/testflows/rbac/requirements/requirements.md) and I've updated the [README.md](https://github.com/ClickHouse/ClickHouse/blob/a03d1d138126548883f5acc636abf2a21e17d94f/tests/testflows/README.md#running-only-selected-tests) to include an example of how to run only the selected tests.

vitlibar commented:
It seems the checks are failed because you're expecting a wrong exit code. The correct error code in that case is `493`, see [ErrorCodes.cpp](https://github.com/ClickHouse/ClickHouse/blob/cf505a92efc0a2a73d6e7c7a456f2b559b7a96bb/src/Common/ErrorCodes.cpp#L461).

vzakaznikov committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into testflows_rbac_syntax_tests

vzakaznikov commented:
> It seems the checks are failed because you're expecting a wrong exit code. The correct error code in that case is `493`, see [ErrorCodes.cpp](https://github.com/ClickHouse/ClickHouse/blob/cf505a92efc0a2a73d6e7c7a456f2b559b7a96bb/src/Common/ErrorCodes.cpp#L461).

Thanks! I will take a look and fix failing tests.

vzakaznikov committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into testflows_rbac_syntax_tests

Commit:
Updating expected error messages to fix failing tests.

vzakaznikov commented:
@vitlibar, `Testflows check` is now green. Could we now approve this PR for merge?
